ports/esp32/machine_bitstream.c

```diff
@@ -91,100 +91,101 @@ static void IRAM_ATTR machine_bitstream_high_low_bitbang(mp_hal_pin_obj_t pin, u
-    bitstream_high_low_1.duration1 = (counter_clk_khz * timing_ns[3]) / 1e6;
-
-    // Install the bits->highlow translator.
-    rmt_translator_init(config.channel, bitstream_high_low_rmt_adapter);
-
-    // Stream the byte data using the translator.
-    check_esp_err(rmt_write_sample(config.channel, buf, len, true));
-
-    // Wait 50% longer than we expect (if every bit takes the maximum time).
-    uint32_t timeout_ms = (3 * len / 2) * (1 + (8 * MAX(timing_ns[0] + timing_ns[1], timing_ns[2] + timing_ns[3])) / 1000);
-    check_esp_err(rmt_wait_tx_done(config.channel, pdMS_TO_TICKS(timeout_ms)));
-
-    // Uninstall the driver.
-    check_esp_err(rmt_driver_uninstall(config.channel));
+    // Example: 500ns = 40000 * 500 / 1e6 = 20 ticks
+    // 20 ticks / 40MHz = 500e-9
+    rmt_bytes_encoder_config_t bytes_encoder_config = {
+        .bit0 = {
+            .level0 = 1,
+            .duration0 = (counter_clk_khz * timing_ns[0]) / 1e6,
+            .level1 = 0,
+            .duration1 = (counter_clk_khz * timing_ns[1]) / 1e6,
+        },
+        .bit1 = {
+            .level0 = 1,
+            .duration0 = (counter_clk_khz * timing_ns[2]) / 1e6,
+            .level1 = 0,
+            .duration1 = (counter_clk_khz * timing_ns[3]) / 1e6,
+        },
+        .flags.msb_first = 1
+    };
+
+    // Install the bits->highlow encoder.
+    rmt_encoder_handle_t encoder;
+    check_esp_err(rmt_new_bytes_encoder(&bytes_encoder_config, &encoder));
+
+    rmt_transmit_config_t tx_config = {
+        .loop_count = 0,
+        .flags.eot_level = 0,
+    };
+
+    // Stream the byte data using the encoder.
+    rmt_encoder_reset(encoder);
+    check_esp_err(rmt_transmit(channel, encoder, buf, len, &tx_config));
+
+    // Wait until completion.
+    rmt_tx_wait_all_done(channel, -1);
+
+    // Disable and release channel.
+    check_esp_err(rmt_del_encoder(encoder));
+    rmt_disable(channel);
+    #if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 3, 0)
+    channel->del(channel);
+    #else
+    rmt_del_channel(channel);
+    #endif
 
     // Cancel RMT output to GPIO pin.
     esp_rom_gpio_connect_out_signal(pin, SIG_GPIO_OUT_IDX, false, false);
+
+    return true;
 }
-#endif
+
+#endif // SOC_RMT_SUPPORTED
+
 /******************************************************************************/
 // Interface to machine.bitstream
 
 void machine_bitstream_high_low(mp_hal_pin_obj_t pin, uint32_t *timing_ns, const uint8_t *buf, size_t len) {
     #if SOC_RMT_SUPPORTED
-    if (esp32_rmt_bitstream_channel_id >= 0) {
-        machine_bitstream_high_low_rmt(pin, timing_ns, buf, len, esp32_rmt_bitstream_channel_id);
-        return;
+    // Falls back to bitbang if rmt_new_tx_channel() fails.
+    // Other ESP errors in RMT version still raise an exception.
+    if (!esp32_rmt_bitstream_enabled || !machine_bitstream_high_low_rmt(pin, timing_ns, buf, len)) {
+        machine_bitstream_high_low_bitbang(pin, timing_ns, buf, len);
     }
-    #endif
+    #else
```

Comment:
Minor nitpick, but this could be written in a way that reduces code duplication of the call to the bitbang function:
```c
 void machine_bitstream_high_low(mp_hal_pin_obj_t pin, uint32_t *timing_ns, const uint8_t *buf, size_t len) {
    #if SOC_RMT_SUPPORTED
    if (esp32_rmt_bitstream_enabled && machine_bitstream_high_low_rmt(pin, timing_ns, buf, len)) {
        // Use of RMT was successful.
        return;
    }
    #endif
    machine_bitstream_high_low_bitbang(pin, timing_ns, buf, len);
}
```
That will also make the diff smaller :smile:

Reply:
Good catch, fixed by https://github.com/micropython/micropython/pull/16293/commits/dd860ef764372d899dbd1d5408ee6e72d78115dd.